C# Basics Workshop (#406)
NuevoFoundation/workshops: content/english/csharp-basics/methods.md

---
title: "Methods"
description: "Introduce methods in C#."
date: 2023-09-13T00:00:00Z
weight: 7
---

## What is a Method?

> So far, we've learned about how to store data in variables, print data and sentences, and make decisions with if-statement, for-loop, and while-loop. 

Reviewer: ```suggestion
So far, we've learned about how to store data in variables, print data and sentences, and make decisions with if-statements, for-loops, and while-loops. 
```

We want to create a ✨ magic box ✨ in code that does the calculation for us no matter which 2 numbers we want to create a sum for!
<img src="../images/method.png" height="250" alt="Num 1 and Num 2 with arrow pointing into a circle with the words Magic Box and an arrow pointing out of the circle pointing to Sum of Num 1 to Num 2"/> 

In C#, a **method** is like the magic box that performs a specific task by running a block of code that can use user inputs.

There are 2 parts in a method: signature and body:

{{% notice note %}}
### Method Signature

To define a method, we need to first write out its <b>method signature</b>. A signature header has four main parts:

**Part Name** | **Description** | **Examples or possible options**
----|----|----
**access specifier** | provides the level of access to the method  | `public` tells the computer that anyone can use this method. `private` tells the computer it can only be called within a class. (We'll learn about classes on the next page!) `protected` tells the computer it can be called by objects of the same class.

> **return type** | data type that is returned to the calling function, technically the return type is not a part of the signaure in C#  | `string` or `int`, use `void` if the method doesn't return anything`.

Reviewer: ```suggestion
**return type** | data type that is returned to the calling function, technically the return type is not a part of the signaure in C#  | `string` or `int`, use `void` if the method doesn't return anything.
```

**method name** | name of the method used to call it | a descriptive name you choose based on what the method does.

> **list of parameters** | list of inputs that must be provided when the method is used | can have zero or more parameters in the form of (`type` `input name`, `type` `input name`, ... ).  Use () for no parameters.

Reviewer: ```suggestion
**list of parameters** | list of inputs that must be provided when the method is used | can have zero or more parameters in the form of (`type` `input name`, `type` `input name`, ... ). Use () for no parameters.
```

**Note:** Once you `return`, nothing else after that in the method runs.

This is an example of how to define a method that sum of numbers from `numA` through `numB`:

```c#
/* Method Name: sumNum
 * Input/Parameter: 2 numbers of type int
 * Funtionality: return the sum from numA through numB
 */
public int sumNum(int numA, int numB){
    int total = 0;
    int num = numA;      // we access the first input with the name numA
    while(num <= numB){  // we access the first input with the name numB
        total = total + num;
        num = num + 1;
    }
    return total; // return statement
}
```

### Method Call

Lastly, to execute a method in your code, we need to make a **method call**. We write out the method name with appropriate input. 
For example:

```C#
sumNum(1, 3); // one line of code that calls method sumNum() with the return value 6
```

Knowing `sumNum(1, 3)` returns an `int` with the sum from 1 to 3, we can do the following to store and then print the value:

To call this method you could do this
```c#
    int sum = 0;
    sum = sumNum(1, 3);
    Console.WrlineLine(sum); // prints 6
```

> One of the resons methods are powerful is we can call them more than once:

Reviewer: ```suggestion
One of the reasons methods are powerful is we can call them more than once:
```
